README.md
# Crispy Cpp Doom
[![Crispy Doom Icon](https://www.chocolate-doom.org/wiki/images/b/be/Crispy-doom.png)](https://github.com/fabiangreffrath/crispy-doom)

[![Top Language](https://img.shields.io/github/languages/top/fabiangreffrath/crispy-doom.svg?style=flat)](https://github.com/fabiangreffrath/crispy-doom)
[![Code Size](https://img.shields.io/github/languages/code-size/fabiangreffrath/crispy-doom.svg?style=flat)](https://github.com/fabiangreffrath/crispy-doom)
[![License](https://img.shields.io/github/license/fabiangreffrath/crispy-doom.svg?style=flat&logo=gnu)](https://github.com/fabiangreffrath/crispy-doom/blob/master/COPYING.md)
[![Release](https://img.shields.io/github/release/fabiangreffrath/crispy-doom.svg?style=flat)](https://github.com/fabiangreffrath/crispy-doom/releases)
[![Release Date](https://img.shields.io/github/release-date/fabiangreffrath/crispy-doom.svg?style=flat)](https://github.com/fabiangreffrath/crispy-doom/releases)
[![Downloads](https://img.shields.io/github/downloads/fabiangreffrath/crispy-doom/latest/total.svg?style=flat)](https://github.com/fabiangreffrath/crispy-doom/releases)
[![Commits](https://img.shields.io/github/commits-since/fabiangreffrath/crispy-doom/latest.svg?style=flat)](https://github.com/fabiangreffrath/crispy-doom/commits/master)
[![Last Commit](https://img.shields.io/github/last-commit/fabiangreffrath/crispy-doom.svg?style=flat)](https://github.com/fabiangreffrath/crispy-doom/commits/master)
[![Travis Build Status](https://img.shields.io/travis/com/fabiangreffrath/crispy-doom.svg?style=flat&logo=travis)](https://travis-ci.com/fabiangreffrath/crispy-doom/)

Crispy Doom is a limit-removing enhanced-resolution Doom source port based on [Chocolate Doom](https://www.chocolate-doom.org/wiki/index.php/Chocolate_Doom).

Its name means that its internal 640x400 resolution looks "crisp" and is also a [slight reference](http://www.mathsisfun.com/recipie.html) to its origin.

## Synopsis

Crispy Doom is a friendly fork of [Chocolate Doom](https://www.chocolate-doom.org/wiki/index.php/Chocolate_Doom) that provides a higher display resolution, removes the [static limits](https://doomwiki.org/wiki/Static_limits) of the Doom engine and offers further optional visual, tactical and physical enhancements while remaining entirely config file, savegame, netplay and demo compatible with the original.

## Objectives and features

Crispy Doom is a source port that aims to provide a faithful Doom gaming experience while also featuring some user-requested improvements and enhancements. It is forked off of Chocolate Doom to take advantage of its free and open-source code base, portability, accuracy and compatibility with Vanilla Doom.

Its core features are:

 * Enhanced 640x400 display resolution, with the original 320x200 resolution still available in the "High Resolution Rendering: Off" mode.
 * Widescreen rendering for using all the available horizontal space of screens with aspect ratios up to 24:9.
 * Uncapped rendering framerate with interpolation and optional vertical synchronization (VSync) with the screen refresh rate.
 * Intermediate gamma correction levels (0.5, 1.5, 2.5 and 3.5).
 * Removal of all static engine limits, or at least raising of the less crucial ones.
 * Full support for the "Doom Classic" WADs shipped with the "Doom 3: BFG Edition", especially the "No Rest For The Living" episode shipped in the NERVE.WAD file.
 * Support for all versions of John Romero's Episode 5: Sigil for Ultimate Doom.

Furthermore, the following optional user-visible and audible features are available:

 * Jumping.
 * Free vertical looking, including mouse look and vertical aiming.
 * Aiming support by a crosshair that may get directly rendered into the game world.
 * A new minimal Crispy HUD, displaying only the status bar numbers.
 * Clean Screenshot feature, enabling to take screenshots without HUD elements and even without status bar numbers and weapon sprites at higher screen sizes.
 * Colorized status bar numbers, HUD texts and blood sprites for certain monsters.
 * Translucency for certain sprites and status bar elements in the Crispy HUD.
 * Randomly mirrored death animations and corpse sprites.
 * Command line options to allow for playing with flipped player weapon sprites and/or entirely flipped level geometry.
 * Players may walk over or under monsters and hanging corpses.
 * Centered Weapons when firing, weapon recoil thrust and pitch.
 * Reports whenever a secret is revealed.
 * Level statistics and extended coloring in the Automap.
 * Playing sounds in full length, and misc. other sound fixes.
 * Demo recording and/or playback timers and progress bar.
 * Demo continue and take-over features, handing controls over to the player when demo playback is finished or interrupted.

Most of these features are disabled by default and need to get enabled either in the in-game "Crispness" menu, in the crispy-doom-setup tool or as command line parameters. They are implemented in a way that preserves demo-compatibility with Vanilla Doom and network game compatibility with Chocolate Doom. Furthermore, Crispy Doom's savegames and config files are compatible, though not identical (see the [Compatibility section in the Wiki](https://github.com/fabiangreffrath/crispy-doom/wiki/Compatibility)), to Vanilla Doom's.

Crispy Doom strives for maximum compatibility with all "limit-removing Vanilla" maps -- but not Boom or ZDoom maps. More specifically, Crispy Doom supports some select advanced features such as [ANIMATED](https://doomwiki.org/wiki/ANIMATED) and [SWITCHES](https://doomwiki.org/wiki/SWITCHES) lumps, MBF sky transfers, SMMU swirling flats and [MUSINFO](https://doomwiki.org/wiki/MUSINFO) -- but neither generalized linedef and sector types nor DECORATE and MAPINFO.

Many additional less user-visible features have been implemented, e.g. fixed engine limitations and crashes, fixed rendering bugs, fixed harmless game logic bugs, full support for DEHACKED files and lumps in BEX format, additional and improved cheat codes, an improved Automap, and many more! Due to the extra DEHACKED states added from [MBF](https://doomwiki.org/wiki/MBF), Crispy Doom supports [enhancer](https://www.doomworld.com/forum/topic/84859-black-ops-smooth-weapons-dehacked-mod) [mods](https://www.doomworld.com/forum/topic/85991-smoothed-smooth-monsters-for-doom-retro-and-crispy-doom) that can make the gameplay even more pleasing to the eyes. For a detailed list of features and changes please refer to the release notes below.

### New controls (with default bindings)

 * Move Forward (alt.) <kbd>W</kbd>
 * Move Backward (alt.) <kbd>S</kbd>
 * Strafe Left (alt.) <kbd>A</kbd>
 * Strafe Right (alt.) <kbd>D</kbd>
 * Jump (bindable to joystick and mouse buttons as well) <kbd>/</kbd> (as in Hexen and Strife)
 * Quick Reverse (bindable to mouse buttons as well)
 * Mouse Look (bindable to mouse buttons or permanent)
 * Look up (bindable to joystick axes as well) <kbd>PgDn</kbd> (as in Heretic)
 * Look down (bindable to joystick axes as well) <kbd>Del</kbd> (as in Heretic)
 * Center view <kbd>End</kbd> (as in Heretic)
 * Toggle always run <kbd>&#8682;</kbd>
 * Toggle vertical mouse movement (new in 5.4)
 * Delete savegame <kbd>Del</kbd>
 * Go to next level
 * Reload current level
 * Save a clean screenshot
 * Toggle Automap overlay mode <kbd>O</kbd>
 * Toggle Automap rotate mode <kbd>R</kbd>
 * Resurrect from savegame (single player mode only) "Run" + "Use"

### New command line parameters

 * `-dm3` specifies the Deathmatch 3.0 rules (weapons stay, items respawn) for netgames (since 4.1).
 * `-episode 1` launches Hell on Earth and `-episode 2` launches No Rest for the Living episode if the Doom 2 IWAD shipped with the Doom 3: BFG Edition is used.
 * `-warp 1a` warps to the secret level E1M10: Sewers of XBox Doom IWAD (since 2.3).
 * `-mergedump <file>` merges the PWAD file(s) given on the command line with the IWAD file and writes the resulting data into the `<file>` given as argument. May be considered as a replacement for the `DEUSF.EXE` tool (since 2.3).
 * `-lumpdump` dumps raw content of a lump into a file (since 5.7).
 * `-blockmap` forces a (re-)building of the BLOCKMAP lumps for loaded maps (since 2.3).
 * `-playdemo demoname -warp N` plays back fast-forward up to the requested map (since 3.0).
 * `-loadgame N -record demoname` and `-loadgame N -playdemo demoname` allow to record and play demos starting from a savegame instead of the level start (since 4.0).
 * `-playdemo demoname1 -record demoname2` plays back fast-forward until the end of demoname1 and continues recording as demoname2 (new in 5.5).
 * `-fliplevels` loads mirrored versions of the maps (this was the default on April 1st up to version 5.0).
 * `-flipweapons` flips the player's weapons (new in 5.3).

### New cheat codes

 * `TNTWEAP` followed by a weapon number gives or removes this weapon (8 = Chainsaw, 9 = SSG). `TNTWEAP0` takes away all weapons and ammo except for the pistol and 50 bullets. Try to load Doom 1 with `DOOM2.WAD` as a PWAD and type `TNTWEAP9` to play the SSG in Doom 1.
 * `TNTEM`, `KILLEM` or `FHHALL` kill all monsters on the current map (and disables all cube spitters).
 * `SPECHITS` triggers all [Linedef actions](https://doomwiki.org/wiki/Linedef_type) on a map at once, no matter if they are enabled by pushing, walking over or shooting or whether they require a key or not. It also triggers all boss monster and Commander Keen actions if possible.
 * `NOTARGET` or `FHSHH` toggle deaf and blind monsters that do not act until attacked.
 * `TNTHOM` toggles the flashing [HOM](https://doomwiki.org/wiki/Hall_of_mirrors_effect) indicator (disabled by default).
 * `SHOWFPS` or `IDRATE` toggle printing the FPS in the upper right corner.
 * `NOMOMENTUM` toggles a debug aid for pixel-perfect positioning on a map (not recommended to use in-game).
 * `GOOBERS` triggers an easter egg, i.e. an "homage to an old friend". ;-)
 * `IDBEHOLD0` disables all currently active power-ups (since 2.2).
 * `IDCLEV00` restarts the current level (since 2.0).
 * `IDMUS00` restarts the current music (new in 5.1).
 * `VERSION` shows the engine version, build date and SDL version (new in 5.1).
 * `SKILL` shows the current skill level (new in 5.5.2).

## Download

Binaries for Windows XP / Vista / 7 / 8.1 / 10 (32-bit binaries compatible with both x86 and x64 editions) are available here:
https://github.com/fabiangreffrath/crispy-doom/releases/download/crispy-doom-5.7.2/crispy-doom-5.7.2-win32.zip

Daily builds of Crispy Doom can be found here:
http://latest.chocolate-doom.org/

Crispy Doom can play nearly all variants of Doom. If you don't own any, you may download the [Shareware version of Doom](http://cdn.debian.net/debian/pool/non-free/d/doom-wad-shareware/doom-wad-shareware_1.9.fixed.orig.tar.gz), extract it and copy the DOOM1.WAD file into your Crispy Doom directory. Alternatively, you may want to play Crispy Doom with [Freedoom](https://www.chocolate-doom.org/wiki/index.php/Freedoom) and a MegaWAD.

### Sources
[![Open Hub](https://www.openhub.net/p/crispy-doom/widgets/project_thin_badge?style=flat&format=gif)](https://www.openhub.net/p/crispy-doom)

The Crispy Doom source code is available at GitHub: https://github.com/fabiangreffrath/crispy-doom.
It can be [downloaded in either ZIP or TAR.GZ format](https://github.com/fabiangreffrath/crispy-doom/releases) 
or cloned via

```
 git clone https://github.com/fabiangreffrath/crispy-doom.git
```

Brief instructions to set up a build system on Windows can be found [in the Crispy Doom Wiki](https://github.com/fabiangreffrath/crispy-doom/wiki/Building-on-Windows). A much more detailed guide is provided [in the Chocolate Doom Wiki](https://www.chocolate-doom.org/wiki/index.php/Building_Chocolate_Doom_on_Windows), but applies to Crispy Doom as well for most parts.

Compilation on Debian systems (Debian 10 "buster" or later) should be as simple as

```
 sudo apt install build-essential automake git
 sudo apt build-dep crispy-doom
```

to install the prerequisites and then

```
 cd crispy-doom
 autoreconf -fiv
 ./configure
 make
```

After successful compilation the resulting binaries can be found in the `src/` directory.

## News

### Crispy Doom 5.7.2

Crispy Doom 5.7.2 has been released on March 13, 2020 to introduce the Compact HUD for widescreen rendering mode and bring some general improvements and fixes.

**Features**

 * A new "Compact HUD" mode has been introduced for widescreen rendering mode which leaves the status bar widgets at the same position as in regular rendering mode instead of moving them to the edges of the screen. The latter is now available in the "Wide HUD" mode.

**Improvements**

 * Compatibility of extended savegames has been improved across forks which change the project name away from "crispy-doom". Zodomaniac noticed this issue with his So Doom fork.
 * No 'Oof' sounds are played anymore upon attempt to manipulate an inactive Crispness menu item, on Zodomaniac's remark.

**Bug Fixes**

 * HUD widgets are now re-initialized immediately after graphics initialization to mitigate the alignment issues that JNechaevsky and Zodomaniac experienced on launching a map from command line.
 * Previous colorization of "percent" status bar widgets is now remembered. This applies to the health and armor widget and forces them to update not only if their value changes, but also if their colorization changes, e.g. when entering the IDDQD cheat. Thanks to unerxai for the report.
 * Horizontal location of the "PAUSE" graphic in widescreen mode has been fixed, thanks to JNechaevsky for spotting this.
 * The widescreen rendering setting is not changed anymore by disabling aspect ratio correction, requested by Zodomaniac.
 * An undetected merge conflict has been fixed in `hexen/r_things.c`, thanks to NickZ for the bug report.

Crispy Doom 5.7.2 is based on Chocolate Doom 3.0.0 and has merged all changes to the Chocolate Doom master branch up to commit [`f1fa7faf`](https://github.com/chocolate-doom/chocolate-doom/commit/f1fa7faf566e3e325c1b1fbf8ace8d5ac2db2b1e).

### Crispy Doom 5.7.1

Crispy Doom 5.7.1 has been released on March 03, 2020 to fix some bugs related to widescreen rendering.

**Improvements**

* In order to improve the discoverability of the "A Secret is Revealed!" feature, its corresponding Crispness menu item has been renamed to "Report Revealed Secrets", based on a discussion with oprypin.
* Widescreen rendering is now disabled in case of disabled aspect ratio correction, as Zodomaniac proved it made no sense.

**Bug Fixes**

* A crash when switching to high resolution or widescreen rendering during a screen wipe has been fixed, reported and investigated by Zodomaniac and JNechaevsky.
* The obtrusive MAP/WAD automap text widget is not drawn anymore in widescreen mode, thanks to plumsinus.
* Horizontal coordinates of automap markers in widescreen mode have been fixed, thanks to unerxai from Doomworld forums.

Crispy Doom 5.7.1 is based on Chocolate Doom 3.0.0 and has merged all changes to the Chocolate Doom master branch up to commit [`f1fa7faf`](https://github.com/chocolate-doom/chocolate-doom/commit/f1fa7faf566e3e325c1b1fbf8ace8d5ac2db2b1e).

### Crispy Doom 5.7 

Crispy Doom 5.7 has been released on February 21, 2020. This release brings proper widescreen rendering as well as other improvements requested by the speed-running community.

**Features**

* Widescreen rendering has been implemented, thanks to JNechaevsky and ice44 for spotting the bugs that had been surfacing in the progress.
* Loading 16 bit RIFF wavs in .wads has been enabled, contributed by SmileTheory.
* `-lumpdump` command line parameter has been added that dumps raw content of a lump into a file.

**Improvements**

* `IDDT` cheat is now reset when re-starting map during demo recording, thanks to Looper for the suggestion.
* Time in demo timer widget is now printed in centiseconds instead of tics. Thanks to Looper and ZeroMaster010 for the proposal.

**Bug Fixes**

* Fix for GUS emulation in presence of midiproc.exe has been pulled from Chocolate Doom, whereto it was contributed by JNechaevsky.

Crispy Doom 5.7 is based on Chocolate Doom 3.0.0 and has merged all changes to the Chocolate Doom master branch up to commit [`f1fa7faf`](https://github.com/chocolate-doom/chocolate-doom/commit/f1fa7faf566e3e325c1b1fbf8ace8d5ac2db2b1e).

### Crispy Doom 5.6.4 

Crispy Doom 5.6.4 has been released on January 14, 2020. This release addresses the community feedback received after 5.6.3.

**Features**

 * Toggling full screen using Enter on num. keyboard is now possible (inherited from Chocolate Doom).
 * `IDCLEV` is now allowed during demo playback and warps to the requested map (starting a new demo).
 * If a new game is started (current level is reloaded/next level is loaded on pressing the respective key) during demo recording, a new demo is started. Thanks to Looper from Doomworld forums for the feature request.
 * 'Intermediate Crispy HUD' without the status bar but with the face and its background in place has been introduced.

**Improvements**

 * `leveltime` is now shown in the demo timer widget during recording instead of the accumulated demo time, thanks to Looper from Doomworld forums for the input.

NEWS.md
## 3.1.0 (2019-??-??)

### General
  * WAD file autoloading was added - WAD and DEH files can be copied into
    an autoload folder to be automatically included on every game start.
  * Music pack configuration has been significantly simplified. By simply
    copying .flac/.ogg music files into a folder they will be automatically
    detected by filename and used.
  * Music packs can now be used with OPL as a fallback, and music pack
    config files can have any name ending in '.cfg'.
  * Network synchronization now uses a PID controller by default, which
    makes games more smooth and more stable, especially for Internet play.
  * UDP hole punching is now used to make servers behind NAT gateways
    automatically accessible to the Internet.
  * OPL emulation now uses Nuked OPL3 v1.8 (thanks nukeykt)
  * The setup tool now uses "Romero Blue" as a background (see the wiki:
    <https://doomwiki.org/wiki/Romero_Blue> for more info).
  * The 0 and 5 keys on the number pad can now be bound independently
    of any other keyboard key (thanks BlooD2ool).
  * With aspect ratio correction disabled, the game can scale to any
    arbitrary size and remove all black borders in full screen mode.
    (thanks chungy)
  * The executable's location is now checked when looking for IWADs.
  * The IWAD files installed by Steam-on-Linux are now detected (thanks
    chungy).
  * It's now possible to use `-response` to load response files.
  * Default savegame name now includes the WAD filename (thanks Fabian).

### Refactorings
  * CMake project files have been added, replacing the Microsoft Visual
    Studio and Code::Blocks files. CMake maintains support for multiple
    IDEs and versions thereof, and reduces developer overhead when updating
    Chocolate Doom (huge thanks to AlexMax for this work).
  * Source code has been retrofitted to fix many compiler warnings and
    add const annotations to many variables (thanks turol).
  * Several functions have been hardened against incomplete reads and
    error conditions, and made safer (thanks turol).
  * Man page generation has been reworked to use autoconf macro
    substitution, making it eaiser for downstream forks to change the
    project name (thanks Jon).
  * We now print a meaningful error message when a savegame cannot be
    loaded (thanks Zodomaniac, chungy).
  * There's now a log file feature for the network code to aid in tracking
    down multiplayer bugs.

### Bug fixes
  * Fixed an exception thrown by the Windows kernel when debugging with
    GDB (thanks AXDOOMER).
  * Loop metadata now works properly with music packs on Windows.
  * Mouse movement is ignored when the game window isn't active (thanks
    Julia Nechaevskaya).
  * A bug was fixed where music would not play after pausing on an
    intermission screen (thanks Julia Nechaevskaya).
  * Timeouts when connecting to a network server were fixed (thanks
    @bradc6).
  * A long-standing bug where some visplane overflows caused crashes was
    fixed (thanks Mike Francis).
  * A multiplayer deadlock bug where clients would stop sending tics after
    missing tics from the server was fixed. There are both client- and
    server- side fixes to fix the problem when playing with older versions
    (thanks MadDog and Mortrixs for help tracking this down).

### Doom
  * Map33 intermission screen and map33-map35 automap names are
    emulated (thanks CapnClever).
  * We now exit gracefully when player starts are missing (thanks Mike
    Francis).
  * We now exit gracefully on levels with a boss brain and no boss spitter
    things (thanks Jason Benaim).
  * It's now possible to play multiplayer with gameversion=1.2.

### Heretic
  * P\_FindNextHighestFloor was changed to match vanilla behavior (thanks
    AXDOOMER).
  * WAD hash table is now generated for speed (thanks Mike Francis).
  * HHE level name replacements now apply on the intermission screen
    (thanks ETTiNGRiNDER).

### Hexen
  * ACS code has been hardened against potential security vulnerabilities.
  * WAD hash table is now generated for speed (thanks AXDOOMER).

### Strife
  * Sehacked replacements of the "empty slot" string now work.
  * VOICES.WAD is now found in a case-insensitive way (thanks Mike Francis).

## 3.0.0 (2017-12-30)

  Chocolate Doom 3.0 is a new major revision. The main change is that
  the codebase has been ported to SDL 2.0. This brings a number of
  benefits, although there have also been some other minor changes (all
  listed below).

  Huge thanks go to the entire Chocolate Doom team for working on the
  port to SDL2, and to all the testers who have found and reported bugs
  during its development.

### General
  * All screen scaling is now performed in hardware, meaning that the
    game can run in arbitrary window sizes in high quality. It can also
    scale to very large resolutions sizes without using large amounts of
    CPU or suffering degraded performance (thanks Fabian).
  * It is now possible to switch between windowed and full screen modes
    while the game is running by pressing alt + enter (thanks Jon)
  * Windows binaries now ship with several previously-optional DLLs.
    This means it is now possible to take PNG screenshots and to use
    digital music packs (FLAC/Ogg Vorbis formats).
  * The game now remembers your preferred monitor and will start on the
    same monitor you were using the last time you played. Windows appear
    centered on the screen.
  * The OS X launcher was tweaked somewhat, and now uses proper path
    controls for choosing files. FreeDM was added as an IWAD.
  * Configuration files on Mac OS X and Unix are now stored in locations
    compliant with the XDG standard (thanks chungy):
    - On Unix: `~/.local/share/chocolate-doom/`
    - On OS X: `~/Library/Application Support/chocolate-doom/`
  * Icons when the game is running are now a higher resolution.
  * Keyboard input is improved and uses the new SDL input API; on

src/sha1.cpp
/* sha1.c - SHA1 hash function
 *	Copyright (C) 1998, 1999, 2000, 2001 Free Software Foundation, Inc.
 *
 * Please see below for more legal information!
 *
 * This file is part of GnuPG.
 *
 * GnuPG is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * GnuPG is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,
 * USA.
 */


/*  Test vectors:
 *
 *  "abc"
 *  A999 3E36 4706 816A BA3E  2571 7850 C26C 9CD0 D89D
 *
 *  "abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq"
 *  8498 3E44 1C3B D26E BAAE  4AA1 F951 29E5 E546 70F1
 */

#include <cstring>

#include "sha1.hpp"

void SHA1_Init(sha1_context_t *hd)
{
    hd->h0      = 0x67452301;
    hd->h1      = 0xefcdab89;
    hd->h2      = 0x98badcfe;
    hd->h3      = 0x10325476;
    hd->h4      = 0xc3d2e1f0;
    hd->nblocks = 0;
    hd->count   = 0;
}

#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wconversion"
/****************
 * Transform the message X which consists of 16 32-bit-words
 */
static void Transform(sha1_context_t *hd, uint8_t *data)
{
    uint32_t a, b, c, d, e, tm;
    uint32_t x[16];

    /* get values from the chaining vars */
    a = hd->h0;
    b = hd->h1;
    c = hd->h2;
    d = hd->h3;
    e = hd->h4;

#ifdef SYS_BIG_ENDIAN
    std::memcpy(x, data, 64);
#else
    {
        int   i;
        uint8_t *p2;
        for (i = 0, p2 = reinterpret_cast<uint8_t *>(x); i < 16; i++, p2 += 4)
        {
            p2[3] = *data++;
            p2[2] = *data++;
            p2[1] = *data++;
            p2[0] = *data++;
        }
    }
#endif


#define K1          0x5A827999L
#define K2          0x6ED9EBA1L
#define K3          0x8F1BBCDCL
#define K4          0xCA62C1D6L
#define F1(x, y, z) (z ^ (x & (y ^ z)))
#define F2(x, y, z) (x ^ y ^ z)
#define F3(x, y, z) ((x & y) | (z & (x | y)))
#define F4(x, y, z) (x ^ y ^ z)

#define rol(x, n) (((x) << (n)) | ((x) >> (32 - (n))))

#define M(i) (tm = x[i & 0x0f] ^ x[(i - 14) & 0x0f]         \
                   ^ x[(i - 8) & 0x0f] ^ x[(i - 3) & 0x0f], \
    (x[i & 0x0f] = rol(tm, 1)))

#define R(a, b, c, d, e, f, k, m) \
    do                            \
    {                             \
        e += rol(a, 5)            \
             + f(b, c, d)         \
             + k                  \
             + m;                 \
        b = rol(b, 30);           \
    } while (0)
    R(a, b, c, d, e, F1, K1, x[0]);
    R(e, a, b, c, d, F1, K1, x[1]);
    R(d, e, a, b, c, F1, K1, x[2]);
    R(c, d, e, a, b, F1, K1, x[3]);
    R(b, c, d, e, a, F1, K1, x[4]);
    R(a, b, c, d, e, F1, K1, x[5]);
    R(e, a, b, c, d, F1, K1, x[6]);
    R(d, e, a, b, c, F1, K1, x[7]);
    R(c, d, e, a, b, F1, K1, x[8]);
    R(b, c, d, e, a, F1, K1, x[9]);
    R(a, b, c, d, e, F1, K1, x[10]);
    R(e, a, b, c, d, F1, K1, x[11]);
    R(d, e, a, b, c, F1, K1, x[12]);
    R(c, d, e, a, b, F1, K1, x[13]);
    R(b, c, d, e, a, F1, K1, x[14]);
    R(a, b, c, d, e, F1, K1, x[15]);
    R(e, a, b, c, d, F1, K1, M(16));
    R(d, e, a, b, c, F1, K1, M(17));
    R(c, d, e, a, b, F1, K1, M(18));
    R(b, c, d, e, a, F1, K1, M(19));
    R(a, b, c, d, e, F2, K2, M(20));
    R(e, a, b, c, d, F2, K2, M(21));
    R(d, e, a, b, c, F2, K2, M(22));
    R(c, d, e, a, b, F2, K2, M(23));
    R(b, c, d, e, a, F2, K2, M(24));
    R(a, b, c, d, e, F2, K2, M(25));
    R(e, a, b, c, d, F2, K2, M(26));
    R(d, e, a, b, c, F2, K2, M(27));
    R(c, d, e, a, b, F2, K2, M(28));
    R(b, c, d, e, a, F2, K2, M(29));
    R(a, b, c, d, e, F2, K2, M(30));
    R(e, a, b, c, d, F2, K2, M(31));
    R(d, e, a, b, c, F2, K2, M(32));
    R(c, d, e, a, b, F2, K2, M(33));
    R(b, c, d, e, a, F2, K2, M(34));
    R(a, b, c, d, e, F2, K2, M(35));
    R(e, a, b, c, d, F2, K2, M(36));
    R(d, e, a, b, c, F2, K2, M(37));
    R(c, d, e, a, b, F2, K2, M(38));
    R(b, c, d, e, a, F2, K2, M(39));
    R(a, b, c, d, e, F3, K3, M(40));
    R(e, a, b, c, d, F3, K3, M(41));
    R(d, e, a, b, c, F3, K3, M(42));
    R(c, d, e, a, b, F3, K3, M(43));
    R(b, c, d, e, a, F3, K3, M(44));
    R(a, b, c, d, e, F3, K3, M(45));
    R(e, a, b, c, d, F3, K3, M(46));
    R(d, e, a, b, c, F3, K3, M(47));
    R(c, d, e, a, b, F3, K3, M(48));
    R(b, c, d, e, a, F3, K3, M(49));
    R(a, b, c, d, e, F3, K3, M(50));
    R(e, a, b, c, d, F3, K3, M(51));
    R(d, e, a, b, c, F3, K3, M(52));
    R(c, d, e, a, b, F3, K3, M(53));
    R(b, c, d, e, a, F3, K3, M(54));
    R(a, b, c, d, e, F3, K3, M(55));
    R(e, a, b, c, d, F3, K3, M(56));
    R(d, e, a, b, c, F3, K3, M(57));
    R(c, d, e, a, b, F3, K3, M(58));
    R(b, c, d, e, a, F3, K3, M(59));
    R(a, b, c, d, e, F4, K4, M(60));
    R(e, a, b, c, d, F4, K4, M(61));
    R(d, e, a, b, c, F4, K4, M(62));
    R(c, d, e, a, b, F4, K4, M(63));
    R(b, c, d, e, a, F4, K4, M(64));
    R(a, b, c, d, e, F4, K4, M(65));
    R(e, a, b, c, d, F4, K4, M(66));
    R(d, e, a, b, c, F4, K4, M(67));
    R(c, d, e, a, b, F4, K4, M(68));
    R(b, c, d, e, a, F4, K4, M(69));
    R(a, b, c, d, e, F4, K4, M(70));
    R(e, a, b, c, d, F4, K4, M(71));
    R(d, e, a, b, c, F4, K4, M(72));
    R(c, d, e, a, b, F4, K4, M(73));
    R(b, c, d, e, a, F4, K4, M(74));
    R(a, b, c, d, e, F4, K4, M(75));
    R(e, a, b, c, d, F4, K4, M(76));
    R(d, e, a, b, c, F4, K4, M(77));
    R(c, d, e, a, b, F4, K4, M(78));
    R(b, c, d, e, a, F4, K4, M(79));

    /* update chainig vars */
    hd->h0 += a;
    hd->h1 += b;
    hd->h2 += c;
    hd->h3 += d;
    hd->h4 += e;
}
#pragma GCC diagnostic pop

/* Update the message digest with the contents
 * of INBUF with length INLEN.
 */
void SHA1_Update(sha1_context_t *hd, uint8_t *inbuf, size_t inlen)
{
    if (hd->count == 64)
    {
        /* flush the buffer */
        Transform(hd, hd->buf);
        hd->count = 0;
        hd->nblocks++;
    }
    if (!inbuf)
        return;
    if (hd->count)
    {
        for (; inlen && hd->count < 64; inlen--)
            hd->buf[hd->count++] = *inbuf++;
        SHA1_Update(hd, nullptr, 0);
        if (!inlen)
            return;
    }

    while (inlen >= 64)
    {
        Transform(hd, inbuf);
        hd->count = 0;
        hd->nblocks++;
        inlen -= 64;
        inbuf += 64;
    }
    for (; inlen && hd->count < 64; inlen--)
        hd->buf[hd->count++] = *inbuf++;
}

#ifdef SYS_BIG_ENDIAN
static uint8_t *write_uint32_big_endian(const uint32_t data, uint8_t *p)
{
    *(uint32_t *)p = data;
    p += 4;
    return p;
}
#else
static uint8_t *write_uint32_little_endian(const uint32_t data, uint8_t *p)
{
    *p++ = static_cast<uint8_t>(data >> 24);
    *p++ = static_cast<uint8_t>(data >> 16);
    *p++ = static_cast<uint8_t>(data >> 8);
    *p++ = static_cast<uint8_t>(data);
    return p;
}
#endif

/* The routine final terminates the computation and
 * returns the digest.
 * The handle is prepared for a new cycle, but adding bytes to the
 * handle will the destroy the returned buffer.
 * Returns: 20 bytes representing the digest.
 */

void SHA1_Final(sha1_digest_t digest, sha1_context_t *hd)
{
    uint32_t t, msb, lsb;
    uint8_t *p;

    SHA1_Update(hd, nullptr, 0); /* flush */
    ;

    t = hd->nblocks;
    /* multiply by 64 to make a byte count */
    lsb = t << 6;
    msb = t >> 26;
    /* add the count */
    t = lsb;
    if ((lsb += static_cast<unsigned int>(hd->count)) < t)
        msb++;
    /* multiply by 8 to make a bit count */
    t = lsb;
    lsb <<= 3;
    msb <<= 3;
    msb |= t >> 29;

    if (hd->count < 56)
    {
        /* enough room */
        hd->buf[hd->count++] = 0x80; /* pad */
        while (hd->count < 56)
            hd->buf[hd->count++] = 0; /* pad */
    }
    else
    {
        /* need one extra block */
        hd->buf[hd->count++] = 0x80; /* pad character */
        while (hd->count < 64)
            hd->buf[hd->count++] = 0;
        SHA1_Update(hd, nullptr, 0); /* flush */
        ;
        std::memset(hd->buf, 0, 56); /* fill next block with zeroes */
    }
    /* append the 64 bit count */
    hd->buf[56] = static_cast<uint8_t>(msb >> 24);
    hd->buf[57] = static_cast<uint8_t>(msb >> 16);
    hd->buf[58] = static_cast<uint8_t>(msb >> 8);
    hd->buf[59] = static_cast<uint8_t>(msb);
    hd->buf[60] = static_cast<uint8_t>(lsb >> 24);
    hd->buf[61] = static_cast<uint8_t>(lsb >> 16);
    hd->buf[62] = static_cast<uint8_t>(lsb >> 8);
    hd->buf[63] = static_cast<uint8_t>(lsb);
    Transform(hd, hd->buf);

    p = hd->buf;

#ifdef SYS_BIG_ENDIAN
    p = write_uint32_big_endian(hd->h0, p);
    p = write_uint32_big_endian(hd->h1, p);
    p = write_uint32_big_endian(hd->h2, p);
    p = write_uint32_big_endian(hd->h3, p);
    p = write_uint32_big_endian(hd->h4, p);
#else
    p = write_uint32_little_endian(hd->h0, p);
    p = write_uint32_little_endian(hd->h1, p);
    p = write_uint32_little_endian(hd->h2, p);
    p = write_uint32_little_endian(hd->h3, p);
    p = write_uint32_little_endian(hd->h4, p);
#endif

    std::memcpy(digest, hd->buf, sizeof(sha1_digest_t));
}

void SHA1_UpdateInt32(sha1_context_t *context, unsigned int val)
{
    uint8_t buf[4];

    buf[0] = static_cast<uint8_t>((val >> 24) & 0xff);
    buf[1] = (val >> 16) & 0xff;
    buf[2] = (val >> 8) & 0xff;
    buf[3] = val & 0xff;

    SHA1_Update(context, buf, 4);
}

void SHA1_UpdateString(sha1_context_t *context, char *str)
{
    SHA1_Update(context, reinterpret_cast<uint8_t *>(str), strlen(str) + 1);
}
